### DISCUSSION TITLE
Questions about the enabling software CSMA / Ca

### DISCUSSION THREAD

**Discussion by tuopu** [openthread/openthread#7759] [Q&A] [answered]
When using Linux platform simulation, test the software CSMA / Ca function, and the radio function enables ot_ RADIO_ CAPS_ TRANSMIT_ RETRIES |OT_ RADIO_ CAPS_ CSMA_ Backoff function also enables opentread_ CONFIG_ MAC_ SOFTWARE_ CSMA_ BACKOFF_ Enable, but no call to get RSSI or energy scanning function is found. Is the implementation based on the RSSI received in the receive function?

**Comment by jwhui** [comment, marked as answer]
OpenThread requires the radio to implement Clear Channel Assessment (CCA). As a result, OpenThread itself does not sample the RSSI to determine whether or not to transmit a given message.

- `OT_RADIO_CAPS_CSMA_BACKOFF` indicates that the radio implements the CSMA backoff.
- `OT_RADIO_CAPS_TRANSMIT_RETRIES` indicates that the radio implements retransmissions.

**Comment by tuopu** [comment]
In other words, opentread itself does not have CCA function. When it needs to be transplanted, it is implemented by the user according to specific hardware. OPENTHREAD_ CONFIG_ MAC_ SOFTWARE_ CSMA_ BACKOFF_ ENABLE
This definition tells openthead protocol stack that the user has realized CCA function. Is that right?
 otPlatRadioGetRssi(otInstance *aInstance)   and otPlatRadioEnergyScanDone(sInstance, rssi); What are the functions of these two functions in thead implementation? Is it independent of CCA；

**Reply by jwhui** [reply]
`OPENTHREAD_ CONFIG_ MAC_ SOFTWARE_ CSMA_ BACKOFF_ ENABLE` only controls whether OpenThread implements the CSMA backoff. It is independent of the CCA and the radio platform must implement the CCA.

`otPlatRadioGetRssi()` and `otPlatRadioEnergyScanDone()` are used to implement features like [Energy Scan](https://openthread.io/reference/group/api-link-link#otlinkenergyscan) and [Channel Monitor](https://openthread.io/reference/group/api-channel-monitor) and are not used to implement CCA.

**Comment by tuopu** [comment]
Which function should be implemented to tell openthead to back off。

**Reply by jwhui** [reply]
If the radio transmission fails due to a CCA failure, then it should pass `OT_ERROR_CHANNEL_ACCESS_FAILURE` via `otPlatRadioTxDone()`: https://github.com/openthread/openthread/blob/3c4caed8a42e08dc1796d4d7c12e383c8a2527c2/include/openthread/platform/radio.h#L808-L824

BTW, you can keep this conversation in a a single thread by replying in the "Write a reply" box, rather than the "Suggest an answer" box.

**Reply by tuopu** [reply]
I understand. After reading other protocol stacks, I realized the CCA algorithm by obtaining the air RSSI. I take it for granted that thread is also implemented in this way.
Thead actually interacts with the user through the otplatradiotxdone function. The user informs thead of what happened, not judging by the hardware state.
